packages/core/src/logger/logger.ts

```diff
@@ -1,144 +1,196 @@
-import { AsyncLocalStorage } from 'async_hooks';
-import { QueueManager } from '../queues';
-import { LogLevel } from '../types';
-import { PluginId } from '../plugins';
-import { LoggerInterface } from './loggerInterface';
-
-export class Logger implements LoggerInterface {
-  private readonly queueManager: QueueManager;
-
-  private readonly globalLogger?: LoggerInterface;
+/* eslint-disable guard-for-in */
+import { v4 } from 'node-uuid';
 
-  private readonly context: AsyncLocalStorage<string>;
-
-  constructor(queueManager: QueueManager, globalLogger?: LoggerInterface) {
-    this.queueManager = queueManager;
-    this.globalLogger = globalLogger;
-    this.context = new AsyncLocalStorage();
+import { AsyncLocalStorage } from 'async_hooks';
+import { LogLevel, TContextBase, LogEntry } from './types';
+import { LogHandler } from './handler';
+
+export class Logger<
+  TContext extends TContextBase = { processId: string },
+  TGlobalContext extends Record<string, any> = {},
+> {
+  private readonly globalContext: TGlobalContext;
+
+  private readonly logHandlers: LogHandler<Partial<TContext>, TGlobalContext>[];
+
+  private readonly asyncStorage: AsyncLocalStorage<Partial<TContext>>;
+
+  public constructor(
+    logHandlers: LogHandler<Partial<TContext>, TGlobalContext>[],
+    globalContext: TGlobalContext,
+  ) {
+    this.globalContext = globalContext;
+    this.logHandlers = logHandlers;
+    this.asyncStorage = new AsyncLocalStorage();
   }
 
-  ensureFork<R>(callback: () => R): R {
-    const tag = this.context.getStore();
+  public ensureFork<R>(callback: () => R): R {
+    const context = this.asyncStorage.getStore();
 
-    if (tag) {
+    if (context) {
       return callback();
     } else {
       return this.fork(callback);
     }
   }
 
-  fork<R>(callback: () => R): R;
-  fork<R>(force: true, callback: () => R): R;
-  fork<R>(callbackOrForce: boolean | (() => R), maybeCallback?: () => R): R {
-    const [callback, force] =
-      typeof callbackOrForce === 'boolean'
-        ? [maybeCallback!, callbackOrForce]
-        : [callbackOrForce, false];
+  public fork<R>(callback: () => R, force?: boolean): R {
```

Comment:
tohle je sice hezčí na tomhle místě, ale představ si, jak se to bude volat, když chceš předat ten parametr - vzhledem k tomu, že callback bude často nějaká inline arrow funkce, budeš mít něco jako:

```typescript
logger.fork(() => {
  // nějakej kód
}, true);
```

ale to ti linter přepíše na:

```typescript
logger.fork(
  () => {
    // nějakej kód
  },
  true,
);
```

srovnej s původní variantou:

```typescript
logger.fork(true, () => {
  // nějakej kód
});
```

proto často preferuju mít některý optional argumenty na začátku argument listu místo na konci - když některej argument často potřebuje víc řádků, je lepší, když je to poslední argument, protože tam, kde se to volá, je to pak přehlednější

Reply:
Ja znam klasickou koncenci, ze na optional argumenty jsou na konci, tak si dodrzuju a tohle mi realne neprislo jako issue

Reply:
v JS se optional leading arguments používaj celkem běžně, možná je to pohrobek doby, kdy nebylo promise api a všechno se řešilo callbacky a callback se vždycky předával jako poslední.. jako asi je to drobnost no..